tortoise/fixture_test.go

```diff
@@ -0,0 +1,791 @@
+}
+
+type beaconAction struct {
+	epoch  uint32
+	beacon types.Beacon
+}
+
+func (b *beaconAction) String() string {
+	return fmt.Sprintf("beacon %d / %v", b.epoch, b.beacon)
+}
+
+func (*beaconAction) deps() []action {
+	return nil
+}
+
+func (b *beaconAction) execute(trt *Tortoise) {
+	trt.OnBeacon(types.EpochID(b.epoch), b.beacon)
+}
+
+// beacon accepts publish epoch and value of the beacon.
+func (s *session) beacon(epoch uint32, value string) *beaconAction {
+	beacon := &beaconAction{epoch: epoch + 1}
+	copy(beacon.beacon[:], value)
+	s.register(beacon)
+	return beacon
+}
+
+type results struct {
+	results []result.Layer
+}
+
+func (r *results) next(lid int) *results {
+	r.results = append(r.results, result.Layer{
+		Layer:  types.GetEffectiveGenesis() + types.LayerID(lid),
+		Blocks: []result.Block{},
+	})
+	return r
+}
+
+func (r *results) verified(lid int) *results {
+	r.results = append(r.results, result.Layer{
+		Layer:    types.GetEffectiveGenesis() + types.LayerID(lid),
+		Verified: true,
+		Blocks:   []result.Block{},
+	})
+	return r
+}
+
+const (
+	hare = 1 << iota
+	valid
+	invalid
+	data
+	local
+)
+
+func (r *results) block(id string, height uint64, fields uint) *results {
+	rst := &r.results[len(r.results)-1]
+	block := result.Block{
+		Valid:   fields&valid > 0,
+		Invalid: fields&invalid > 0,
+		Hare:    fields&hare > 0,
+		Data:    fields&data > 0,
+		Local:   fields&local > 0,
+	}
+	copy(block.Header.ID[:], id)
+	block.Header.LayerID = rst.Layer
+	block.Header.Height = height
+	rst.Blocks = append(rst.Blocks, block)
+	return r
+}
+
+type updateActions struct {
+	tb      testing.TB
+	expect  []result.Layer
+	actions []action
+}
+
+func (u *updateActions) String() string {
+	return "updates"
+}
+
+func (u *updateActions) deps() []action {
+	return u.actions
+}
+
+func (u *updateActions) execute(trt *Tortoise) {
+	u.tb.Helper()
+	updates := trt.Updates()
+	for i := range updates {
+		// TODO(dshulyak) don't know yet how to implement
+		updates[i].Opinion = types.Hash32{}
+	}
+	require.Equal(u.tb, u.expect, updates)
+}
+
+func (s *session) updates(tb testing.TB, expect *results) {
+	deps := make([]action, len(s.actions))
+	copy(deps, s.actions)
+	s.register(&updateActions{tb, expect.results, deps})
+}
+
+func (s *session) runInorder() {
+	s.runActions(s.actions)
+}
+
+func (s *session) runActions(actions []action) {
+	trt := s.tortoise()
+	for _, a := range actions {
+		a.execute(trt)
+	}
+}
+
+func (s *session) runRandomTopoN(n int) {
+	for i := 0; i < n; i++ {
+		s.runRandomTopo()
+	}
+}
+
+func (s *session) runRandomTopo() {
+	// TODO(dshulyak) compute and execute all topological paths
+	// or optionally a subset of them if all takes too long
+	rst := []action{}
+	empty := []action{}
+	forward := map[action]map[action]struct{}{}
+	reverse := map[action]map[action]struct{}{}
+	for _, act := range s.actions {
+		for _, dep := range act.deps() {
+			if reverse[act] == nil {
+				reverse[act] = map[action]struct{}{}
+			}
```

Comment:
move to the outer loop?

Reply:
realized that reverse should have no key is act.deps() is empty. moving to outer loop breaks it